lib/std/builtin.zig

```diff
@@ -160,54 +160,341 @@ pub const OptimizeMode = enum {
+    /// Deprecated; use `.kernel`.
+    pub const Kernel: CallingConvention = .kernel;
+    /// Deprecated; use `.spirv_fragment`.
+    pub const Fragment: CallingConvention = .spirv_fragment;
+    /// Deprecated; use `.spirv_vertex`.
+    pub const Vertex: CallingConvention = .spirv_vertex;
+
+    /// The default Zig calling convention when neither `export` nor `inline` is specified.
+    /// This calling convention makes no guarantees about stack alignment, registers, etc.
+    /// It can only be used within this Zig compilation unit.
+    auto,
+
+    /// The calling convention of a function that can be called with `async` syntax. An `async` call
+    /// of a runtime-known function must target a function with this calling convention.
+    /// Comptime-known functions with other calling conventions may be coerced to this one.
+    @"async",
+
+    /// Functions with this calling convention have no prologue or epilogue, making the function
+    /// uncallable in regular Zig code. This can be useful when integrating with assembly.
+    naked,
+
+    /// This calling convention is exactly equivalent to using the `inline` keyword on a function
+    /// definition. This function will be semantically inlined by the Zig compiler at call sites.
+    /// Pointers to inline functions are comptime-only.
+    @"inline",
+
+    // Calling conventions for the x86_64 architecture.
+    x86_64_sysv: CommonOptions,
+    x86_64_win: CommonOptions,
+    x86_64_regcall_v3_sysv: CommonOptions,
+    x86_64_regcall_v4_win: CommonOptions,
+    x86_64_vectorcall: CommonOptions,
+    x86_64_interrupt: CommonOptions,
+
+    // Calling conventions for the x86 architecture.
+    x86_sysv: X86RegparmOptions,
+    x86_win: X86RegparmOptions,
+    x86_stdcall: X86RegparmOptions,
+    x86_fastcall: CommonOptions,
+    x86_thiscall: CommonOptions,
+    x86_thiscall_mingw: CommonOptions,
+    x86_regcall_v3: CommonOptions,
+    x86_regcall_v4_win: CommonOptions,
+    x86_vectorcall: CommonOptions,
+    x86_interrupt: CommonOptions,
+
+    // Calling conventions for the aarch64 architecture.
+    aarch64_aapcs: CommonOptions,
+    aarch64_aapcs_darwin: CommonOptions,
+    aarch64_aapcs_win: CommonOptions,
+    aarch64_vfabi: CommonOptions,
+    aarch64_vfabi_sve: CommonOptions,
+
+    // Calling convetions for the arm architecture.
     /// ARM Procedure Call Standard (obsolete)
-    /// ARM-only.
-    APCS,
-    /// ARM Architecture Procedure Call Standard (current standard)
-    /// ARM-only.
-    AAPCS,
+    arm_apcs: CommonOptions,
+    /// ARM Architecture Procedure Call Standard
+    arm_aapcs: CommonOptions,
     /// ARM Architecture Procedure Call Standard Vector Floating-Point
-    /// ARM-only.
-    AAPCSVFP,
-    /// x86-64-only.
-    SysV,
-    /// x86-64-only.
-    Win64,
-    /// AMD GPU, NVPTX, or SPIR-V kernel
-    Kernel,
-    // Vulkan-only
-    Fragment,
-    Vertex,
+    arm_aapcs_vfp: CommonOptions,
+    arm_aapcs16_vfp: CommonOptions,
```

Comment:
I forget, did we decide to keep these (despite `arm_aapcs` vs `arm_aapcs_vfp` just being soft vs hard float) specifically because C compilers let you pick between them?

If yes, then `defaultCCallingConvention()` is getting this wrong, and needs to pick based on `target.abi.floatAbi()`.

Reply:
I actually don't know anything about these, I think I copied them from you without really looking into it. (That doesn't necessarily mean we didn't talk about it, I'm forgetful!)

Could you remind me of the situation here?

Reply:
Looks like the relevant discussion was here: https://zsf.zulipchat.com/#narrow/stream/454360-compiler/topic/callconv.20stuff/near/468038539

So yeah, looks like that was indeed the reason it stayed, in which case `defaultCCallingConvention()` should just be fixed to check the float ABI and pick the right one.

Reply:
Just to check, this would be correct in `cCallingConvention`?
```zig
        .arm, .armeb, .thumb, .thumbeb => switch (target.abi.floatAbi()) {
            .soft => .{ .arm_aapcs = .{} },
            .hard => .{ .arm_aapcs_vfp = .{} },
        },
```

Reply:
Yep, that's correct. (`arm_aapcs16_vfp` is dead since it was only ever used by Apple on their watchOS devices running 32-bit Arm.)

Reply:
@mlugg actually, sorry, correction:

```zig
        .arm, .armeb, .thumb, .thumbeb => if (target.os.tag == .netbsd) .arm_apcs else switch (target.abi.floatAbi()) {
            .soft => .{ .arm_aapcs = .{} },
            .hard => .{ .arm_aapcs_vfp = .{} },
        },
```

Apparently NetBSD is still living in the aforementioned Before Times. :shrug: